x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/Realms.java

```diff
@@ -70,7 +74,9 @@ public Realms(Settings settings, Environment env, Map<String, Realm.Factory> fac
         this.threadContext = threadContext;
         this.reservedRealm = reservedRealm;
         assert factories.get(ReservedRealm.TYPE) == null;
-        this.realms = initRealms();
+        final List<RealmConfig> realmConfigs = buildRealmConfigs();
+        disabledNativeRealmTypes = findDisabledNativeRealmTypes(realmConfigs);
```

Comment:
Because the logic in `Realms` is already very complicated, I would prefer we don't create another field like `disabledNativeRealmTypes`.

I think you can move the logic from `findDisabledNativeRealmTypes` inside `initRealms`, because this part further down
```
for (List<Realm> realmList : Arrays.asList(standardRealms, nativeRealms)) {
    if (realmList.isEmpty()) {
            if (realmList.isEmpty()) {
                addNativeRealms(realmList);
            }
     }
     assert realmList.contains(reservedRealm) == false;
     realmList.add(0, reservedRealm);
     assert realmList.get(0) == reservedRealm;
}
```
is now redundant, and is dangerous because it can create *new* realms for realms that we only expect one to exist.

Reply:
It's always great to simplify the code. I have changed the code based on your suggestion. One minor thing is that we still need `realmList.add(0, reservedRealm)` from the above code snippet. But I always managed to simplify it a bit. Hopefully it reads better now.